base-trezor-keyring.js

```diff
@@ -0,0 +1,562 @@
+ *
+ * @param {TypedTransaction | OldEthJsTransaction} tx
+ * @returns {tx is OldEthJsTransaction} Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(tx) {
+  return typeof tx.getChainId === 'function';
+}
+
+class BaseTrezorKeyring extends EventEmitter {
+  constructor(opts = {}) {
+    super();
+    this.type = keyringType;
+    this.accounts = [];
+    this.hdk = new HDKey();
+    this.page = 0;
+    this.perPage = 5;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.deserialize(opts);
+  }
+
+  /**
+   * Performs any asynchronous initialisation
+   *
+   * @returns {Promise<void>}
+   */
+  init() {
+    throw new Error('Method init not implemented');
+  }
+
+  /**
+   * Gets the model, if known.
+   * This may be `undefined` if the model hasn't been loaded yet.
+   *
+   * @returns {"T" | "1" | undefined}
+   */
+  getModel() {
+    return this.model;
+  }
+
+  /**
+   * Clears the iframe created by Trezor
+   *
+   * @returns {Promise<void>}
+   */
+  dispose() {
+    throw new Error('Method dispose not implemented');
+  }
+
+  /**
+   * Gets public key from device
+   *
+   * @returns {Promise}
+   */
+  _getPublicKey(_payload) {
+    throw new Error('Method _getPublicKey not implemented');
+  }
+
+  /**
+   * Signs transaction using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTransaction(_payload) {
+    throw new Error('Method _ethereumSignTransaction not implemented');
+  }
+
+  /**
+   * Signs message using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignMessage(_payload) {
+    throw new Error('Method _ethereumSignMessage not implemented');
+  }
+
+  /**
+   * Signs typed data using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTypedData(_payload) {
+    throw new Error('Method _ethereumSignTypedData not implemented');
+  }
+
+  serialize() {
+    return Promise.resolve({
+      hdPath: this.hdPath,
+      accounts: this.accounts,
+      page: this.page,
+      paths: this.paths,
+      perPage: this.perPage,
+      unlockedAccount: this.unlockedAccount,
+    });
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
+    this.accounts = opts.accounts || [];
+    this.page = opts.page || 0;
+    this.perPage = opts.perPage || 5;
+    return Promise.resolve();
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk && this.hdk.publicKey);
+  }
+
+  unlock() {
+    if (this.isUnlocked()) {
+      return Promise.resolve('already unlocked');
+    }
+    return new Promise((resolve, reject) => {
+      this._getPublicKey({
+        path: this.hdPath,
+        coin: 'ETH',
+      })
+        .then((response) => {
+          if (response.success) {
+            this.hdk.publicKey = Buffer.from(response.payload.publicKey, 'hex');
+            this.hdk.chainCode = Buffer.from(response.payload.chainCode, 'hex');
+            resolve('just unlocked');
```

Comment:
Can these strings be made into named constants so they can be referenced to in handling by calling code?

Reply:
I didn't even realise how that string was being used. I'll definitely refactor that even if it's only to add the options as a constants (an enum-like in js). Too easy to make mistakes otherwise.